django_saml2_auth/utils.py

```diff
@@ -0,0 +1,577 @@
+from django.shortcuts import render
+from django.urls import NoReverseMatch, reverse
+from django.utils.module_loading import import_string
+from pkg_resources import parse_version
+from saml2 import BINDING_HTTP_POST, BINDING_HTTP_REDIRECT, entity
+from saml2.client import Saml2Client
+from saml2.config import Config as Saml2Config
+from saml2.response import AuthnResponse
+
+from .errors import *
+from .exceptions import SAMLAuthError
+
+
+def run_hook(function_path: str,
+             *args: Optional[Tuple[Any]],
+             **kwargs: Optional[Mapping[str, Any]]) -> Optional[Any]:
+    """Runs a hook function with given args and kwargs. For example, given
+    "models.User.create_new_user", the "create_new_user" function is imported from
+    the "models.User" module and run with args and kwargs.
+
+    Args:
+        function_path (str): A path to a hook function,
+            e.g. models.User.create_new_user (static method)
+
+    Raises:
+        SAMLAuthError: function_path isn't specified
+        SAMLAuthError: There's nothing to import. Check your hook's import path!
+        SAMLAuthError: Import error
+        SAMLAuthError: Re-raise any exception caused by the called function
+
+    Returns:
+        Optional[Any]: Any result returned from running the hook function. None is returned in case
+            of any exceptions, errors in arguments and related issues.
+    """
+    if not function_path:
+        raise SAMLAuthError("function_path isn't specified", extra={
+            "exc_type": ValueError,
+            "error_code": EMPTY_FUNCTION_PATH,
+            "reason": "There was an error processing your request.",
+            "status_code": 500
+        })
+
+    path = function_path.split(".")
+    if len(path) < 2:
+        # Nothing to import
+        raise SAMLAuthError("There's nothing to import. Check your hook's import path!", extra={
+            "exc_type": ValueError,
+            "error_code": PATH_ERROR,
+            "reason": "There was an error processing your request.",
+            "status_code": 500
+        })
+
+    module_path = ".".join(path[:-1])
+    result = None
+    try:
+        cls = import_string(module_path)
+        result = getattr(cls, path[-1])(*args, **kwargs)
+    except ImportError as exc:
+        raise SAMLAuthError(str(exc), extra={
+            "exc": exc,
+            "exc_type": ImportError,
+            "error_code": IMPORT_ERROR,
+            "reason": "There was an error processing your request.",
+            "status_code": 500
+        })
+    except Exception as exc:
+        raise SAMLAuthError(str(exc), extra={
+            "exc": exc,
+            "exc_type": type(exc),
+            "error_code": GENERAL_EXCEPTION,
+            "reason": "There was an error processing your request.",
+            "status_code": 500
+        })
+
+    return result
+
+
+def create_new_user(email: str, firstname: str, lastname: str) -> Type[Model]:
```

Comment:
I don't like having a lot of not directly related functions in `utils.py`, makes it hard to maintain. It looks like you have some saml processing functions, user management and error handling here.

Maybe separate those into something like 
`utils/saml.py`, `utils/user.py`, `utils/error.py`? (names just a suggestion, you probably knoe how to do it better)

Reply:
Thanks for the suggestion. I was also thinking about the same thing and now that you've mentioned it, I'll do it. :+1: